src/HttpRequest.cc

```diff
@@ -754,6 +752,16 @@ HttpRequest::manager(const CbcPointer<ConnStateData> &aMgr, const AccessLogEntry
     }
 }
 
+bool
+HttpRequest::mustGoToOriginalDestination() const
+{
+    // TODO: exclude flags.internal requests
+    // TODO: REQMOD/redirection occur after host header verification
+    const auto isIntercepted = !flags.redirected && masterXaction->hasListeningInterceptedPort();
```

Comment:
> Amos: This is a test about the current request state, not related to the port.

I will try to rephrase Eduard question in hope to clarify his bewilderment (which I share): Why do you this test of the current request state cannot be related to the port that the request was received on? The port configuration does not _fully_ determine the request state, of course, and the expression in question contains factors beyond the port configuration (i.e. flags.redirected), but it is not clear why you think the port is irrelevant to producing the right result here?

@eduard-bagdasaryan, is this a new condition or will it produce decisions identical the existing official code (when proxySurrogateHttps is false)? AFAICT, you are just copying the existing duplicated conditions from PeerSelector::resolveSelected() and FwdState::start(), but please correct me if I am wrong.

Reply:
>  it produce decisions identical the existing official code

These conditions are the same as in master. Moreover, I believe that this is true for other PR's conditions as well (only wrapping these conditions with some helper methods).

Reply:
> These conditions are the same as in master [when proxySurrogateHttps is false].

Thank you. @yadij, please re-evaluate this change request, keeping in mind the the new condition is equivalent to the old condition when proxySurrogateHttps is false.

* If you think the old condition is wrong, perhaps fixing it is outside this PR scope?
* If you think the new condition is wrong, which can only be because of the addition of proxySurrogateHttps to hasListeningInterceptedPort() via interceptedSomewhere() definition, then please resolve this change request. You already have another change request about the interceptedSomewhere() definition, and the issue of interceptedSomewhere() result for true proxySurrogateHttps should be debated [there](https://github.com/squid-cache/squid/pull/835/files#r644453644).

@eduard-bagdasaryan, if Amos disagrees with my suggestion to resolve this change request, and it is still not clear how this change request should be addressed, then I recommend _suspending_ this discussion until that interceptedSomewhere() change [request](https://github.com/squid-cache/squid/pull/835/files#r644453644) is resolved.

Reply:
> I am not sure I understand your request - what is the 'current request state'?

By "current request" I mean the httprequest instance which is the latest/current one representing the transaction. Could be from client, from redirector or adaptation. The port flags can only be guaranteed accurate for the client one.

It simplifies flag checks if request copies the changeable flags. Client getting set from port config. Then each request copy/cloner gets to adjust during the create. Otherwise flags stay cost. 

>  I recommend suspending this discussion until that interceptedSomewhere() change request is resolved.

Fine by me. Same for the other requests like this.

Reply:
> It simplifies flag checks if request copies the changeable flags. Client getting set from port config. Then each request copy/cloner gets to adjust during the create. Otherwise flags stay const.

Yes, this happens for both flags and masterXaction. Official code used HttpRequest::flags. PR code gets the same info from HttpRequest::masterXaction. Both HttpRequest fields are inherited/copied across adaptations.